nixos/tests/zfs-tests.nix

```diff
@@ -0,0 +1,221 @@
+import ./make-test-python.nix ({ lib, pkgs, ... } : let
+  # It's kind of awkward that this is required...
+  # Is there a better way to do this?
+  # We now have to rebuild ZFS.
+  zfsForTesting = pkgs.zfs.overrideAttrs (oldAttrs: {
+    removeTests = false;
+    postPatch = (oldAttrs.postPatch or "") + ''
+      # TODO(ricsch): Is overriding with /run/current-system/sw/bin acceptable?
+      substituteInPlace scripts/zfs-tests.sh \
+          --replace '$STF_PATH/ksh' ${pkgs.ksh}/bin/ksh \
+          --replace 'LOSETUP=' 'LOSETUP=${pkgs.util-linux}/bin/losetup #' \
+          --replace 'export PATH=$STF_PATH' '#' \
+          --replace 'DMSETUP=' 'DMSETUP=${pkgs.lvm2}/bin/dmsetup #' \
+          --replace '$STF_PATH/gzip' '${pkgs.gzip}/bin/gzip' \
+          --replace '$STF_PATH/gunzip' '${pkgs.gzip}/bin/gunzip' \
+          --replace '/sbin/fsck.ext4' '${pkgs.e2fsprogs}/bin/fsck.ext4' \
+          --replace '/sbin/mkfs.ext4' '${pkgs.e2fsprogs}/bin/mkfs.ext4' \
+          --replace '$STF_PATH/awk' '${pkgs.gawk}/bin/awk' \
+          --replace 'SYSTEM_DIRS="/usr/local/bin /usr/local/sbin"' 'SYSTEM_DIRS="/run/current-system/sw/bin"'
+
+      substituteInPlace scripts/common.sh.in \
+          --replace 'export ZTS_DIR=' 'export ZTS_DIR=${placeholder "zfs_tests"}/share/zfs #' \
+          --replace 'export SCRIPT_DIR=' 'export SCRIPT_DIR=${placeholder "zfs_tests"}/share/zfs #' \
+          --replace 'export ZDB=' 'export ZDB=${placeholder "out"}/bin/zdb #' \
+          --replace 'export ZFS=' 'export ZFS=${placeholder "out"}/bin/zfs #' \
+          --replace 'export ZPOOL=' 'export ZPOOL=${placeholder "out"}/bin/zpool #' \
+          --replace 'export ZTEST=' 'export ZTEST=${placeholder "out"}/bin/ztest #' \
+          --replace 'export ZFS_SH=' 'export ZFS_SH=${placeholder "out"}/share/zfs/zfs.sh #'
+
+      # Fix ksh paths in test suite.
+      # patchShebangs doesn't work due to the scripts not being executable.
+      # It doesn't seem logical to make them executable either.
+      find -name "*.ksh" -exec sed -i 's,/bin/ksh,${pkgs.ksh}/bin/ksh,' {} \;
+
+      # Patching maybe required for more binaries
+      # Maybe FHS environment would be better?
+    '';
+    postInstall = (oldAttrs.postInstall or "") + ''
+      # Ensure that the tests are not deleted.
+      # This is a requirement.
+      ls $zfs_tests/share/zfs/zfs-tests || exit 1
+    '';
+  });
```

Comment:
I added it with a includeTests argument, because I can't get it working with a separate output due to problems with them being dependent on each other or something like that.